pyexasol/callback.py

```diff
@@ -74,33 +78,88 @@ def import_from_pandas(pipe, src, **kwargs):
+    src.to_csv(
+        wrapped_pipe,
+        header=False,
+        index=False,
+        lineterminator="\n",
+        quoting=csv.QUOTE_NONNUMERIC,
+        **kwargs,
+    )
 
-    # https://github.com/pandas-dev/pandas/pull/45302
-    if packaging.version.parse(pandas.__version__) >= packaging.version.parse("1.5.0"):
-        return src.to_csv(
-            wrapped_pipe,
-            header=False,
-            index=False,
-            lineterminator="\n",
-            quoting=csv.QUOTE_NONNUMERIC,
-            **kwargs
-        )
-    else:
-        return src.to_csv(
-            wrapped_pipe,
-            header=False,
-            index=False,
-            line_terminator="\n",
-            quoting=csv.QUOTE_NONNUMERIC,
-            **kwargs
+
+def get_parquet_files(source: Union[list[Path], Path, str]) -> list[Path]:
+    if isinstance(source, str):
+        matches = glob.glob(source)
+        return sorted(
+            [filepath for i in matches if (filepath := Path(i)) and filepath.is_file()]
         )
+    if isinstance(source, Path):
+        if source.is_file():
+            return [source]
+        elif source.is_dir():
+            return sorted(source.glob("*.parquet"))
+    if isinstance(source, list):
+        not_a_valid_path_file = [
+            filepath
+            for filepath in source
+            if not isinstance(filepath, Path) or not filepath.is_file()
+        ]
+        if len(not_a_valid_path_file) > 0:
+            raise ValueError(
+                f"source {source} contained entries which were not `Path` or valid `Path` files {not_a_valid_path_file}"
+            )
+        return source
+    raise ValueError(
+        f"source {source} is not a supported type (Union[list[Path], Path, str])."
+    )
+
+
+def import_from_parquet(pipe, source: Union[list[Path], Path, str], **kwargs):
+    """
+    Basic example how to import from pyarrow parquet file(s)
+
+    Args:
+        source: Local filepath specification(s) to process. Can be one of:
+            - list[pathlib.Path]: list of specific files
+            - pathlib.Path: can be either a file or directory. If it's a directory,
+            all files matching this pattern *.parquet will be processed.
+            - str: representing a filepath which already contains a glob pattern
+            (e.g., "/local_dir/*.parquet")
+        **kwargs:
+            Custom params for "pyarrow.csv.WriteOptions"
+
+    Please note that nested or hierarchical types are not supported.
+    """
+    from pyarrow import (
+        csv,
+        parquet,
+        types,
+    )
+
+    if not (parquet_files := get_parquet_files(source)):
+        raise ValueError(f"source {source} does not match any files")
+
+    for file in parquet_files:
+        parquet_file = parquet.ParquetFile(file)
+        num_row_groups = parquet_file.num_row_groups
+
+        for i in range(num_row_groups):
+            row_group_table = parquet_file.read_row_group(i)
```

Comment:
We might need to think about if a user wants to select specific columns. This function can receive a list of columns and would only read these.
Another question, I am asking me now, is if the row group is the correct abstraction, or if we go even smaller there is also iter_batches https://arrow.apache.org/docs/python/generated/pyarrow.parquet.ParquetFile.html#pyarrow.parquet.ParquetFile.iter_batches. However, this would also read a row group, but if a conversion from parquet to arrow is needed, it would avoid allocating memory for the whole row group.

Reply:
Since the initial specifications were for local files, I updated to memory_map=True & using iter_batches.

I switched where **kwargs was passed, which was WriteOptions, to instead go to the iter_batches. Now, columns can be selected & in their desired output order.